docs/proposals/crt_management_proposal.md

```diff
@@ -0,0 +1,113 @@
+# Container Runtime Management
+
+## Description
+origin and openshift-ansible support multiple container runtimes.  This proposal
+is related to refactoring how we handle those runtimes in openshift-ansible.
+
+### Problems addressed
+We currently don't install docker during the install at a point early enough to
+not fail health checks, and we don't have a good story around when/how to do it.
+This is complicated by logic around containerized and non-containerized installs.
+
+A web of dependencies can cause changes to docker that are unintended and has
+resulted in a series of work-around such as 'skip_docker' boolean.
+
+We don't handle docker storage because it's BYO.  By moving docker to a prerequisite
+play, we can tackle storage up front and never have to touch it again.
+
+container_runtime logic is currently spread across 3 roles: docker, openshift_docker,
+and openshift_docker_facts.  The name 'docker' does not accurately portray what
+the role(s) do.
+
+## Rationale
+* Refactor docker (and related meta/fact roles) into 'container_runtime' role.
+* Strip all meta-depends on container runtime out of other roles and plays.
+* Create a 'prerequisites.yml' entry point that will setup various items
+such as container storage and container runtime before executing installation.
+* All other roles and plays should merely consume container runtime, should not
+configure, restart, or change the container runtime as much as feasible.
+
+## Design
+
+The container_runtime role should be comprised of 3 'pseudo-roles' which will be
+consumed using include_role; each component area should be enabled/disabled with
+a boolean value, defaulting to true.
+
+I call them 'pseudo-roles' because they are more or less independent functional
+areas that may share some variables and act on closely related components.  This
+is an effort to reuse as much code as possible, limit role-bloat (we already have
+an abundance of roles), and make things as modular as possible.
+
+```yaml
+# prerequisites.yml
+- include: std_include.yml
+- include: container_runtime_setup.yml
+...
+# container_runtime_setup.yml
+- hosts: "{{ openshift_runtime_manage_hosts | default('oo_nodes_to_config') }}"
```

Comment:
Will this be a new group?

Reply:
What groups need us to manage containers for a user?  Some users might not want us to manage any groups, and then would not need to run these plays at all.  Other users might just need nodes.  Others might want to manage many groups (etcd, load balancers, other things that are a part of their inventory that need to be configured the same as nodes in regard to container runtime).

This is just an example and a point of discussion.  How should we handle what groups we run against?  I don't like the network of booleans we currently have.  I would prefer to default to the nodes group, and if people are using containerized hosts for other roles (etc), they can specify those group names.

Reply:
As long as this group is paired with the booleans required to turn on or off I think we are ok reusing the same group.